ocaml/xapi/updateinfo.ml

```diff
@@ -441,18 +449,79 @@ module Severity = struct
         raise Api_errors.(Server_error (invalid_updateinfo_xml, []))
 end
 
+module GuidanceInUpdateInfo = struct
+  type t = (Guidance.kind * Guidance.t list) list
+
+  let value_of_xml = function
+    | Xml.Element ("value", _, [Xml.PCData v]) ->
+        Some (Guidance.of_string v)
+    | Xml.Element (unexpected, _, _) ->
+        warn "Ignore unexpected guidance value XML tag %s" unexpected ;
+        None
+    | _ ->
+        None
+
+  exception Unsupported_kind_xml
+
+  let kind_of_xml = function
+    | "recommended" ->
+        Guidance.Recommended
+    | "mandatory" ->
+        Guidance.Mandatory
+    | "full" ->
+        Guidance.Full
+    | "livepatch" ->
+        Guidance.Livepatch
+    | _ ->
+        raise Unsupported_kind_xml
+
+  let default =
+    let open Guidance in
+    [(Mandatory, []); (Recommended, []); (Full, []); (Livepatch, [])]
+
+  let of_xml xml_blocks =
+    List.fold_left
+      (fun acc xml_block ->
+        match xml_block with
+        | Xml.Element (kind_xml, _, values_in_xml) -> (
+          match kind_of_xml kind_xml with
+          | kind ->
+              let values = List.filter_map value_of_xml values_in_xml in
+              (kind, values) :: List.remove_assoc kind acc
+          | exception Unsupported_kind_xml ->
+              warn "Unsupported guidance kind XML tag %s" kind_xml ;
+              acc
+        )
+        | _ ->
+            warn "Ignore unexpected XML node in guidance." ;
```

Comment:
Can we log something about the unexpected node, I think then the log will be meaningful.

Reply:
It may not be worth to log a XML node the structure of which is unknown.